Add ability to selectively claim joint interfaces

wxmerkt opened the pull request:
This PR adds the ability to selectively claim joints and thus allows us to have multiple controllers run at the same time, e.g. have IHMC's controller run the effort joints with their whole body controller and us run separate controllers for neck and finger/forearm joints. 

Tested on Val. Sequence to get it to work:
1. start controller manager
2. start this controller using the NeckControl.launch
3. start the IHMC controller (either through command line or through ui_builder)

wxmerkt committed:
Claim and load PositionJointInterface

wxmerkt committed:
More informative start-up message on claimed resources

wxmerkt committed:
Subscribe to PositionJointInterface and append them to EST_ROBOT_STATE, fixes strict=True as well

wxmerkt committed:
Comments

wxmerkt committed:
Add position control

wxmerkt committed:
Minor formatting

wxmerkt committed:
Add neck control launch and config files

wxmerkt committed:
Add check whether a list of joints has been assigned, otherwise assign all

mauricefallon commented:
This badly needs to be functionalised and perhaps the lcm i/o broken out into a separate cpp for readability. The functions inside the namespace shouldn't be indented.

I'd like to do away with having the header in the include directory. We don't use that approach in the rest of the code base

Probably should review this as is and make these changes afterwards

mauricefallon commented on src/LCM2ROSControl.cpp:
```diff
@@ -41,7 +41,18 @@ namespace valkyrie_translator
+        // If we have joints assigned to just us, claim those, otherwise claim all
+        std::vector<std::string> joint_names_;
+        if (!controller_nh.getParam("joints", joint_names_))
+          ROS_INFO_STREAM("Could not get assigned list of joints, will resume to claim all");
```
I don't like splitting the config between ROS params and lcm-config and director-json (and possibly also drake config).
this is the kind of decentralisation that I abhor in ROS.

But for now we will use this.

wxmerkt replied:
Agreed. However for the purpose of a single controller only running on the robot without other dependencies this the preferred and recommended approach. The getParam is read from the yaml file where we specify whether to assign specific joints (e.g. the neck) or whether to grab all available joints (in the case of no joints specified in the parameter). LCM or director json config isn't compatible with ROS Control per se at the moment, but we can look into making it

mauricefallon commented on src/LCM2ROSControl.cpp:
```diff
@@ -204,43 +272,79 @@ namespace valkyrie_translator
         imuchannel << "VAL_IMU_" << iter->first;
         lcm_imu_msg.utime = utime;
-        for (i=0; i<3; i++){
+        for (int i=0; i<3; i++){
```
Iterator 0:3 for quaternion seems to be a bug ... missing the 4th element?

wxmerkt replied:
Good catch - probably a bug, need to check this (in sim we don't get IMU data and on the robot we haven't checked it so far).

mauricefallon replied:
Just looked at the raw file. The 4th element was being done manually after the loop. Please redo that as that without using this loop. Is a pretty sloppy way of avoiding a few lines.

wxmerkt replied:
Agreed, I didn't revise/clean up bits that weren't needed to be touched yet. Will do in the next PR

wxmerkt committed:
Fix CMakeLists.txt:
Previously it was necessary to comment out pods.cmake if compiled in a catkin workspace where another pods.cmake was located in. This nifty solution goes around this bug

wxmerkt committed:
Update header info

wxmerkt committed:
Added gitignore

wxmerkt committed:
Address quaternion comment

wxmerkt committed:
minor cleanup
